### PR TITLE
[npud] Add ContextManager class

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit adds ContextManager to manange Context in npu daemon. It is used by Core class.

Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Draft PR: https://github.com/Samsung/ONE/pull/9989

### CODE REVIEW

File: runtime/service/npud/core/ContextManager.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  // TODO Consider the possibility of reusing the same address.
+  context->contextId = reinterpret_cast<ContextID>(context.get());
+  context->npuContext = npuContext;
+  *contextId = context->contextId;
+  _contexts.emplace_back(std::move(context));
+
+  this->listContexts();
+}
+
+void ContextManager::deleteContext(ContextID contextId)
```
[the remark this answers is not in the document]
Comment (jyoungyun) [reply runtime/service/npud/core/ContextManager.cc:43]: @ragmani 

This function deletes the given context Id. What does it mean to delete from?

[the remark this answers is not in the document]
Comment (ragmani) [reply runtime/service/npud/core/ContextManager.cc:43]: `_contexts.erase(iter, _contexts.end());` removes the elements in the range [`iter`, `_contexts.end()`).

[the remark this answers is not in the document]
Comment (jyoungyun) [reply runtime/service/npud/core/ContextManager.cc:43]: The `remove_if` shifts all elements forward except for the item to remove and returns the location of the object to be erased.  And then it erases the location to the end.

This operation just deletes the given item.
So I don't understand why I have to put `from`.

[the remark this answers is not in the document]
Comment (ragmani) [reply runtime/service/npud/core/ContextManager.cc:43]: Sorry, it's my misunderstood. I didn't know about `remove_if`.

File: runtime/service/npud/core/ContextManager.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+{
+  const auto iter =
+    std::remove_if(_contexts.begin(), _contexts.end(),
+                   [&](std::unique_ptr<Context> &c) { return c->contextId == contextId; });
+  if (iter == _contexts.end())
+  {
+    return;
+  }
```
Comment (ragmani) [inline runtime/service/npud/core/ContextManager.cc:51]: Is this if block necessary?

Comment (jyoungyun) [reply runtime/service/npud/core/ContextManager.cc:51]: Yes. If there is no item in the list, it just returns.

Comment (ragmani) [reply runtime/service/npud/core/ContextManager.cc:51]: Just FYI, the following codes are safe if there is no item in the list.
But, there is no problem if this block exists as it is.

Comment (jyoungyun) [reply runtime/service/npud/core/ContextManager.cc:51]: Thank you for information. However, I want to return it explicitly. Can I keep this code as it is?

File: runtime/service/npud/core/ContextManager.cc [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+{
+
+ContextManager::ContextManager() noexcept {}
+
+ContextManager::~ContextManager() noexcept { _contexts.clear(); }
+
+void ContextManager::newContext(NpuContext *npuContext, ContextID *contextId)
+{
+  auto context = std::make_unique<Context>();
+  // TODO Consider the possibility of reusing the same address.
```
Comment (ragmani) [inline runtime/service/npud/core/ContextManager.cc:34]: I'm not sure about it but I think it depends how you allocate and deallocate `Context` instances.
If there are deallocated heap memory for those instances, there is a possibility of reusing the same address. But, it seems that there is no duplicated addresses in the vector if those deallocated instances were removed from the vector.

**Review by ragmani — APPROVED** [review]
LGTM

**Review by hseok-oh — APPROVED** [review]

**Pull request merged**